asyncpg/protocol/coreproto.pyx

```diff
@@ -796,30 +798,55 @@ cdef class CoreProtocol:
 
         self._send_bind_message(portal_name, stmt_name, bind_data, limit)
 
-    cdef _bind_execute_many(self, str portal_name, str stmt_name,
-                            object bind_data):
-
-        cdef WriteBuffer buf
-
+    cdef _execute_many_init(self):
         self._ensure_connected()
         self._set_state(PROTOCOL_BIND_EXECUTE_MANY)
 
         self.result = None
         self._discard_data = True
-        self._execute_iter = bind_data
-        self._execute_portal_name = portal_name
-        self._execute_stmt_name = stmt_name
 
-        try:
-            buf = <WriteBuffer>next(bind_data)
-        except StopIteration:
+    cdef _execute_many_send(self, object buffers):
+        self._ensure_connected()
+        self.transport.writelines(buffers)
+
+    cdef _execute_many_rollback(self, object error):
+        cdef:
+            WriteBuffer packet
+            WriteBuffer buf
+
+        self._ensure_connected()
+        if self.state != PROTOCOL_CANCELLED:
+            self._set_state(PROTOCOL_BIND_EXECUTE_MANY_ROLLBACK)
+        packet = WriteBuffer.new()
+
+        # We have no idea if we are in an explicit transaction or not,
+        # therefore we raise an exception in the database to mark current
+        # transaction as failed anyway to prevent partial commit
+        buf = self._build_parse_message('',
+                                        'DO language plpgsql $$ BEGIN '
```

Comment:
Because if we are in a subtransaction (savepoint), ROLLBACK would abort through all subtransactions. Okay in this case maybe it is the only way to e.g. hack in SELECT 1/0? Please kindly advise :)

Reply:
I think we should not touch the existing transaction state.  If we are currently in transaction, just do nothing.  An unhandled exception in `executemany` iterator would crash the whole thing, and if they _do_ handle the exception, then the intention is clear and explicit.  We only want to guard against unhandled exceptions.